storage: introduce non-temporal commands

andreimatei opened the pull request:
Introduce a new flag for requests - isNonTemporal. This is set by the lease
requests to specify that they don't need to be serialized through the
command queue (they were using the range's first key, which is a bogus
thing to sync on).

Also sets the isAlone flag on LeaseRequest and LeaseTransfer requests.

cc @tschottdorf

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8310)

<!-- Reviewable:end -->

andreimatei committed:
storage: introduce non-temporal commands

Introduce a new flag for requests - isNonKV. This is set by the lease
requests to specify that they don't need to be serialized through the
command queue (they were using the range's first key, which is a bogus
thing to sync on).

Also sets the isAlone flag on LeaseRequest and LeaseTransfer requests.

tamird commented:
No tests?

On Wed, Aug 3, 2016 at 3:57 PM, Andrei Matei <notifications@github.com>
wrote:

> Introduce a new flag for requests - isNonTemporal. This is set by the lease
> requests to specify that they don't need to be serialized through the
> command queue (they were using the range's first key, which is a bogus
> thing to sync on).
>
> Also sets the isAlone flag on LeaseRequest and LeaseTransfer requests.
>
> cc @tschottdorf <https://github.com/tschottdorf>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/cockroachdb/cockroach/pull/8310
> Commit Summary
>
>    - storage: introduce non-temporal commands
>
> File Changes
>
>    - *M* roachpb/api.go
>    <https://github.com/cockroachdb/cockroach/pull/8310/files#diff-0> (9)
>    - *M* roachpb/batch.go
>    <https://github.com/cockroachdb/cockroach/pull/8310/files#diff-1> (15)
>    - *M* storage/replica.go
>    <https://github.com/cockroachdb/cockroach/pull/8310/files#diff-2> (80)
>
> Patch Links:
>
>    - https://github.com/cockroachdb/cockroach/pull/8310.patch
>    - https://github.com/cockroachdb/cockroach/pull/8310.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8310>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPCB5jrzAr0yLY8MPqssjkTvSQvIQks5qcPKxgaJpZM4JcCNX>
> .
>

andreimatei commented:
Well I'm not really sure how to test the absence of this synchronization. Suggestions?

bdarnell commented:
To test, start a write on the start key of the range, and use a TestingCommandFilter to make it block. Verify that you can still renew the lease while that write is blocking. 

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[roachpb/api.go, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH1mvL4HGhKVklTIQo:-KOH1mvL4HGhKVklTIQp:-1772802818) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/api.go#L86)):*
> ```Go
> 	// Some commands can skip interacting with the command queue and the timestamp
> 	// cache. For example, RequestLeaseRequest is sequenced exclusively by Raft.
> 	isNonTemporal
> ```

`NonKV` might be a better name. In my mind the thing that ties these requests together is that they don't really have a key; they point to the start key of a range to identify it but don't actually touch that key.

---

*[roachpb/api.go, line 822 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH-KsHxuRwUcsh37iM:-KOH-KsHxuRwUcsh37iN:1937747711) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/api.go#L822)):*
> ```Go
> func (*NoopRequest) flags() int               { return isRead } // slightly special
> func (*MergeRequest) flags() int              { return isWrite }
> func (*TruncateLogRequest) flags() int        { return isWrite }
> ```

`TruncateLogRequest`, `ComputeChecksumRequest`, `VerifyChecksumRequest`, and `ChangeFrozenRequest` are also non-temporal. 

---

*[roachpb/batch.go, line 99 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH-pfA043-buv5z9Yd:-KOH-pfB154wA8Pfgynt:663623473) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/batch.go#L99)):*
> ```Go
> // IsSingleRequest returns true iff the BatchRequest contains a single request.
> func (ba *BatchRequest) IsSingleRequest() bool {
> 	return len(ba.Requests) == 0
> ```

Shouldn't that be `== 1`? 

---

*[roachpb/batch.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH0RqacXFvyQZqAiQv:-KOH0RqacXFvyQZqAiQw:1983195271) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/batch.go#L107)):*
> ```Go
> func (ba *BatchRequest) SingleRequestIsNonTemporal() bool {
> 	if !ba.IsSingleRequest() {
> 		panic(fmt.Sprintf("SingleRequestIsNonTemporal called on batch with multiple requests: %s", ba))
> ```

Panicking here means that it's only possible to call this safely immediately after `IsSingleRequest()`, so it's redundant. Either return false if `!IsSingleRequest` so this method can be used on its own, or remove the SingleRequest check so you can call it as `isNonTemporal := ba.IsSingleRequest() && ba.IsNonTemporal()`.

---

*[storage/replica.go, line 1281 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH0kXLiuzfBBxLdLXb:-KOH0kXLiuzfBBxLdLXc:1166623398) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/storage/replica.go#L1281)):*
> ```Go
> 			return nil, roachpb.NewError(err)
> 		}
> 	}
> ```

Add an `else` branch that assigns a no-op function to `endCmdsFunc` so you don't need to repeat the `isNonTemporal` check below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-:-KOH0ugNhBNd1yuc6cOh:70220628)*
<!-- Sent from Reviewable.io -->

andreimatei committed:
storage: send lease requests throught the sender replica

They used to be proposed to Raft directly, to avoid the command queue.
But now the command queue is gone and we can remove this special
direct-proposal code.
I've had to add another cmd flag about skipping the lease check on
proposal (i.e. redirectOnOrAcquireLease). I've made it separate from the
isLease() method on a batch, which just checks for RequestLease, because
LeaseTransfer is not always symetric. E.g. it's still a good idea to
check in Raft that the lease under which LeaseTransfer was proposed is
valid.

andreimatei commented:
So it turns out that lease extensions and transfers were proposed directly to Raft, bypassing the `addWriteCmd()` in replica - which checks, among others, the command queue. But now that we have the flag for skipping the queue, I've added a commit which makes these lease commands be sent in the usual way, through the replica.

The commits to review are `r5,r6`. `r2` is a separate PR - #8352 .

As described in `r6`, I've added another flag lease requests to skip the `redirectOnOrAcquireLease` check and I've left the `isLease` method alone. One thing I'm not sure about is whether the following use of `isLease` should be extended to transfers or not:
https://github.com/cockroachdb/cockroach/blob/master/storage/replica.go#L1425
Os `lastAssignedLeaseIndex` is *not* incremented on LeaseTransfer. I think that's good, because `lastAssignedLeaseIndex` seems to have something to do with commands that can be proposed by anybody, not just by the lease holder (which is not the case for a transfer), but I don't really know...

---

Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[roachpb/api.go, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH1mvL4HGhKVklTIQo:-KOVz018hLxBBcG2p1tI:bjayb1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/api.go#L86)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `NonKV` might be a better name. In my mind the thing that ties these requests together is that they don't really have a key; they point to the start key of a range to identify it but don't actually touch that key.
</details>
I think neither name is suggestive enough that it doesn't require an explanation. Out of the two, I prefer non-temporal... `NonKV` is not suggestive at all to me...

---

*[roachpb/api.go, line 822 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH-KsHxuRwUcsh37iM:-KOVwOsWQK0yFrj5NFZb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/api.go#L822)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `TruncateLogRequest`, `ComputeChecksumRequest`, `VerifyChecksumRequest`, and `ChangeFrozenRequest` are also non-temporal. 
</details>
Done.

---

*[roachpb/batch.go, line 99 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH-pfA043-buv5z9Yd:-KOVwUvgBXBbZ8YXqiXD:b-hv9eb7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/batch.go#L99)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Shouldn't that be `== 1`? 
</details>
yes :)

---

*[roachpb/batch.go, line 107 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH0RqacXFvyQZqAiQv:-KOVwXtpg99n-4tgU9xx:bxwd9gb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/batch.go#L107)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Panicking here means that it's only possible to call this safely immediately after `IsSingleRequest()`, so it's redundant. Either return false if `!IsSingleRequest` so this method can be used on its own, or remove the SingleRequest check so you can call it as `isNonTemporal := ba.IsSingleRequest() && ba.IsNonTemporal()`.
</details>
ok, option 1

---

*[storage/replica.go, line 1281 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH0kXLiuzfBBxLdLXb:-KOVyzGMZ8kwBaJP13ne:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/storage/replica.go#L1281)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Add an `else` branch that assigns a no-op function to `endCmdsFunc` so you don't need to repeat the `isNonTemporal` check below.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-:-KOW3k7OlwL-JJNpVUkh:b7bgcy2)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 1 of 5 files at r2, 2 of 5 files at r3, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[roachpb/api.go, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH1mvL4HGhKVklTIQo:-KOb5dn9O4Tt9f9z_LR8:bdzq9kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/api.go#L86)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think neither name is suggestive enough that it doesn't require an explanation. Out of the two, I prefer non-temporal... `NonKV` is not suggestive at all to me...
</details>
I agree that neither name is very suggestive, but I think `NonKV` is better than `NonTemporal`. The lease commands don't use the timestamp cache, but they are very much "temporal" in that they are about time. The common factor is that they are addressed by a RangeID rather than a key. 

If we find a good name here it might help in the `keys` package as well, where we have `LocalRangeID` keys and `Local`, non-`RangeID` keys. 

---

*[roachpb/api.go, line 831 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb8AjYvhw08WxfudGK:-KOb8AjYvhw08WxfudGL:b-6yrd2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/roachpb/api.go#L831)):*
> ```Go
> }
> func (*TransferLeaseRequest) flags() int {
> 	return isWrite | isAlone | isNonTemporal | skipLeaseCheck
> ```

Don't transfers need the lease?

---

*[roachpb/batch.go, line 58 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb6gcAiU96QXGtQJjG:-KOb6gcAiU96QXGtQJjH:bnayr36) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/roachpb/batch.go#L58)):*
> ```Go
> 
> // IsLease returns whether the batch consists of a single RequestLease request.
> // Note that TransferLease requests return false.
> ```

Explain why: that `RequestLease` is the only command that a non-leaseholder can issue. Maybe rename to `IsLeaseRequest` while you're here. Shouldn't this use the skipLeaseCheck flag now instead of the request type?

If combined with my suggestion below to give `skipLeaseCheck` to `ChangeFrozen`, this would mean that `ChangeFrozen` requests also don't increment the lease applied index. I *think* that's what we want. 

---

*[roachpb/batch.go, line 99 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb8bgWuh7-zabPcUn1:-KOb8bgWuh7-zabPcUn2:b-pyjxul) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/roachpb/batch.go#L99)):*
> ```Go
> 
> // IsSingleRequest returns true iff the BatchRequest contains a single request.
> func (ba *BatchRequest) IsSingleRequest() bool {
> ```

You can unexport this now. 

---

*[roachpb/batch.go, line 105 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb8dcmrU6NtlVRnb6j:-KOb8dcmrU6NtlVRnb6k:b-yaxd9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/roachpb/batch.go#L105)):*
> ```Go
> // IsSingleNonTemporalRequest returns true iff the batch contains a single
> // request, and that request has the non-temporal flag set.
> func (ba *BatchRequest) IsSingleNonTemporalRequest() bool {
> ```

I think I'd remove "Single" from these method names. 

---

*[storage/replica.go, line 1372 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb9QL_hio66iuEWvK4:-KOb9QLakwtVkpoulMDD:b-htxppg) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/storage/replica.go#L1372)):*
> ```Go
> 	if !ba.IsSingleSkipLeaseCheckRequest() {
> 		// This replica must have range lease to process a write, except when it's
> 		// an attempt to unfreeze the Range. These are a special case in which any
> ```

Should `ChangeFrozen` have the `skipLeaseCheck` flag, and then we can remove the special case for `IsFreeze` below? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[roachpb/api.go, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH1mvL4HGhKVklTIQo:-KOeEfKa3kKy_6aq1MqI:b-z8u8ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/api.go#L86)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I agree that neither name is very suggestive, but I think `NonKV` is better than `NonTemporal`. The lease commands don't use the timestamp cache, but they are very much "temporal" in that they are about time. The common factor is that they are addressed by a RangeID rather than a key. 
> 
> If we find a good name here it might help in the `keys` package as well, where we have `LocalRangeID` keys and `Local`, non-`RangeID` keys. 
</details>
`isMeta`? I'm not particularly fond of that. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[roachpb/api.go, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH1mvL4HGhKVklTIQo:-KOea2Ee0ONEf2Y-xmu0:b-5ds31) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/api.go#L86)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `isMeta`? I'm not particularly fond of that. 
</details>
`isMeta` suggests `RangeLookup` (which deals with `meta` keys) to me. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 6 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[roachpb/api.go, line 86 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOH1mvL4HGhKVklTIQo:-KOfp0DvBWHyvc-q_9uP:b103zn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1849590dde934321f6ee31ec22cb8963d9ad00ea/roachpb/api.go#L86)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `isMeta` suggests `RangeLookup` (which deals with `meta` keys) to me. 
</details>
I was using "non-temporal" by analogy with x86 SSE instructions which don't obey the usual load/store ordering. OK, `isNonKV`.

---

*[roachpb/api.go, line 831 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb8AjYvhw08WxfudGK:-KOfpetkKYZBPxW_fzVC:b-mk3lzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/roachpb/api.go#L831)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Don't transfers need the lease?
</details>
It does, and it's specially checked in `AdminTransferLease()` at proposal time and in the usual way for write commands at apply time.
But it can't be checked at propose time through the `redirectOrAcquire` call because, by the time that call is made, the replica has registered that a transfer is in progress and `redirectOrAcquire` already tentatively redirects to the next holder. Added a comment here.

---

*[roachpb/batch.go, line 58 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb6gcAiU96QXGtQJjG:-KOfsIk3qViI9aPSxWb8:b1z1xgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/roachpb/batch.go#L58)):*
Renamed and added that comment.

> Shouldn't this use the skipLeaseCheck flag now instead of the request type?

The thing is that the checks for the lease need to be different at proposal time vs at apply time. At apply time, we want `TransferLease` to be checked. But at proposal time it can't be checked in the usual way,  as explained in `TransferLeaseRequest`'s flags. So at proposal time we use the flag, at apply time we use the request type.

---

*[roachpb/batch.go, line 99 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb8bgWuh7-zabPcUn1:-KOg-0h5D4WP12DVkA5O:b9r8359) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/roachpb/batch.go#L99)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> You can unexport this now. 
</details>
yeah... But I think it's a good enough method to export...

---

*[roachpb/batch.go, line 105 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb8dcmrU6NtlVRnb6j:-KOg-7ivpBA7ki4jTOiH:b-zgbfou) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/roachpb/batch.go#L105)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I think I'd remove "Single" from these method names. 
</details>
well the problem is that I find this `hasFlag` method a bit dangerous because it returns `true` if *any* request in the batch have a flag. And the other batch method use that and don't care about  `single`/non-single. But since this method exceptionally does care, I'd keep `Single` in the name.

---

*[storage/replica.go, line 1372 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-KOb9QL_hio66iuEWvK4:-KOfzPsGZYjE82h4dw8i:b-2pwr2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/81ce25c72e21457b056fc7f4cdae849577253ab4/storage/replica.go#L1372)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Should `ChangeFrozen` have the `skipLeaseCheck` flag, and then we can remove the special case for `IsFreeze` below? 
</details>
as discussed, that'd change behavior, since now we only allow a `ChangeFreeze` to proceed if the range is frozen; otherwise we check the lease in the usual way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 6 files at r7, 4 of 4 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8310#-:-KOkH7u1dOyJGzkojYW_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on roachpb/batch.go:
```diff
@@ -47,16 +47,19 @@ func (ba *BatchRequest) SetActiveTimestamp(nowFn func() hlc.Timestamp) error {
+// IsLeaseRequest returns whether the batch consists of a single RequestLease
+// request. Note that TransferLease requests return false.
+// RequestLease requests are special because they're the only type of requests a
+// non-lease-holder can propose.
```
This description sounds like the `skipLeaseCheck` flag, but that flag is both on `RequestLease` and `TransferLease` (and that in turn is justified inline in the flags on `TransferLeaseRequest`). Some of this confusion should be reducible.
